File: solutions/haskell/twelve-days/1/src/TwelveDays.hs
module TwelveDays (recite) where

recite :: Int -> Int -> [String]
recite start stop = map recite' [start..stop]
    where
        recite' day = text1 <> (days !! (day-1)) <> text2 <> concat (reverse $ take day gifts)

text1 :: String
text1 = "On the "

days :: [String]
days = ["first", "second", "third", "fourth", "fifth", "sixth", "seventh", "eighth", "ninth", "tenth", "eleventh", "twelfth"]

text2 :: String
text2 = " day of Christmas my true love gave to me: "

gifts :: [String]
gifts = ["a Partridge in a Pear Tree.", "two Turtle Doves, and ", "three French Hens, ", "four Calling Birds, ", "five Gold Rings, ", "six Geese-a-Laying, ", "seven Swans-a-Swimming, ", "eight Maids-a-Milking, ", "nine Ladies Dancing, ", "ten Lords-a-Leaping, ", "eleven Pipers Piping, ", "twelve Drummers Drumming, "]

File: solutions/fsharp/twelve-days/1/TwelveDays.fs
// Dedicated to Shree DR.MDD
module TwelveDays

let days =
    [ "first"
      "second"
      "third"
      "fourth"
      "fifth"
      "sixth"
      "seventh"
      "eighth"
      "ninth"
      "tenth"
      "eleventh"
      "twelfth" ]

let gifts =
    [ "two Turtle Doves"
      "three French Hens"
      "four Calling Birds"
      "five Gold Rings"
      "six Geese-a-Laying"
      "seven Swans-a-Swimming"
      "eight Maids-a-Milking"
      "nine Ladies Dancing"
      "ten Lords-a-Leaping"
      "eleven Pipers Piping"
      "twelve Drummers Drumming" ]

let line dayNum =
    let dayWord = days.Item(dayNum - 1)
    let firstGift =
        if dayNum > 1 then "and a Partridge in a Pear Tree" else "a Partridge in a Pear Tree"
    let giftList =
        firstGift
        :: gifts
        |> List.take dayNum
        |> List.rev
        |> String.concat ", "
    sprintf "On the %s day of Christmas my true love gave to me: %s." dayWord giftList

let recite startDay endDay = [ startDay .. endDay ] |> List.map line
